File: app/src/main/java/de/marcelknupfer/studentenausweisguthaben/PopupActivity.java
package de.marcelknupfer.studentenausweisguthaben;

import android.annotation.TargetApi;
import android.app.ActivityOptions;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.nfc.NfcAdapter;
import android.nfc.Tag;
import android.os.Build;
import android.os.Bundle;

import android.util.Log;
import android.util.Pair;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.FragmentManager;
import com.codebutler.farebot.card.desfire.DesfireException;
import de.marcelknupfer.studentenausweisguthaben.cardreader.Readers;
import de.marcelknupfer.studentenausweisguthaben.cardreader.ValueData;

/**
 * Created by wenzel on 28.11.14.
 */
public class PopupActivity extends AppCompatActivity {

	private boolean darkMode = false;
	private SharedPreferences sharedPres;

	private static String TAG = PopupActivity.class.getSimpleName();

	private static final String VALUE_TAG = "value";
	private ValueFragment valueFragment;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		sharedPres = getSharedPreferences(getString(R.string.preference_key), Context.MODE_PRIVATE);
		darkMode = sharedPres.getBoolean(getString(R.string.preference_darkmode_key), false);
		if (darkMode) {
			setTheme(R.style.DarkMode);
		}
		setContentView(R.layout.popup_main);


		Toolbar toolbar = (Toolbar) findViewById(R.id.main_toolbar);
		toolbar.setTitle(R.string.app_name);
		setSupportActionBar(toolbar);

		Log.i(TAG, "activity started");


		FragmentManager fm = getSupportFragmentManager();

		valueFragment = (ValueFragment) fm.findFragmentByTag(VALUE_TAG);
		if (valueFragment ==null) {
			valueFragment = new ValueFragment();
			fm.beginTransaction().replace(R.id.main, valueFragment,VALUE_TAG).commit();
		}



		if (NfcAdapter.ACTION_TECH_DISCOVERED.equals(getIntent().getAction())) {
			Log.i(TAG, "Started by tag discovery");
			onNewIntent(getIntent());
		}
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		menu.add(Menu.NONE, R.id.fullscreen, Menu.NONE, R.string.fullscreen).setIcon(R.drawable.ic_action_full_screen).setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
		return true;
	}

	@Override
	public void onStart() {
		super.onStart();
		if (ValueHolder.data!=null)
			valueFragment.setValueData(ValueHolder.data);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.fullscreen:
				Intent intent = new Intent(PopupActivity.this, MainActivity.class);
				intent.setAction(MainActivity.ACTION_FULLSCREEN);
				intent.putExtra(MainActivity.EXTRA_VALUE, valueFragment.getValueData());

				if (Build.VERSION.SDK_INT>=Build.VERSION_CODES.LOLLIPOP) {
					animateActivity21(intent);
				} else if (Build.VERSION.SDK_INT>=Build.VERSION_CODES.JELLY_BEAN) {
					animateActivity16(intent);
				} else {
					startActivity(intent);
					overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
				}

				return true;
			default:
				return super.onOptionsItemSelected(item);
		}
	}


	@TargetApi(21)
	private void animateActivity21(Intent intent) {
		ActivityOptions options;
		//TODO check what happens if last is empty
		if (valueFragment.getValueData()!=null) {
			options = ActivityOptions.makeSceneTransitionAnimation(PopupActivity.this,
					Pair.create(findViewById(R.id.current), "current"),
					Pair.create(findViewById(R.id.last), "last"),
					Pair.create(findViewById(R.id.main_toolbar), "toolbar")
			);
		} else {
			options = ActivityOptions.makeSceneTransitionAnimation(PopupActivity.this,
					Pair.create(findViewById(R.id.current), "current"),
					Pair.create(findViewById(R.id.main_toolbar), "toolbar")
			);
		}

		startActivity(intent, options.toBundle());
		//ActivityCompat.startActivity(PopupActivity.this,intent,
		//		ActivityOptionsCompat.makeSceneTransitionAnimation(PopupActivity.this).toBundle());
		//finish();
	}

	@TargetApi(16)
	private void animateActivity16(Intent intent) {
		View root = findViewById(R.id.popupRoot);

		ActivityOptions options = ActivityOptions.makeScaleUpAnimation(root,root.getLeft(), root.getTop(), root.getWidth(), root.getHeight());
		startActivity(intent, options.toBundle());
	}

	@Override
	protected void onNewIntent(Intent intent) {
		super.onNewIntent(intent);

		if (NfcAdapter.ACTION_TECH_DISCOVERED.equals(intent.getAction())) {
			Log.i(TAG,"Discovered tag with intent: " + intent);
			Tag tag = intent.getParcelableExtra(NfcAdapter.EXTRA_TAG);


			try {
				ValueData val = Readers.getInstance().readTag(tag);

				valueFragment.setValueData(val);


			} catch (DesfireException e) {
				Toast.makeText(this, R.string.communication_fail, Toast.LENGTH_SHORT).show();
			}
		}
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		valueFragment = null;
	}
}

File: app/src/main/java/de/marcelknupfer/studentenausweisguthaben/ui/activities/PopupActivity.kt
package de.marcelknupfer.studentenausweisguthaben.ui.activities

import android.annotation.TargetApi
import android.app.ActivityOptions
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.os.Build
import android.os.Bundle

import android.util.Log
import android.util.Pair
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast

import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.codebutler.farebot.card.desfire.DesfireException
import de.marcelknupfer.studentenausweisguthaben.R
import de.marcelknupfer.studentenausweisguthaben.cardreader.Readers
import de.marcelknupfer.studentenausweisguthaben.cardreader.ValueHolder
import de.marcelknupfer.studentenausweisguthaben.ui.fragments.ValueFragment
import java.lang.IllegalStateException

/**
 * @author © 2019 Marcel Knupfer
 * Popup Activity that will open when NFC card is recognized
 */
class PopupActivity: AppCompatActivity() {

    private var darkMode = false
    private var sharedPres: SharedPreferences? = null
    private var valueFragment: ValueFragment? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sharedPres = getSharedPreferences(getString(R.string.preference_key), Context.MODE_PRIVATE)
        darkMode = sharedPres!!.getBoolean(getString(R.string.preference_darkmode_key), false)
        if (darkMode) {
            setTheme(R.style.DarkMode)
        }
        setContentView(R.layout.popup_main)


        val toolbar = findViewById<View>(R.id.main_toolbar) as Toolbar
        toolbar.setTitle(R.string.credit)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeAsUpIndicator(R.drawable.ic_close_white_24dp)

        Log.i(TAG, "activity started")


        val fm = supportFragmentManager

        valueFragment = fm.findFragmentByTag(VALUE_TAG) as ValueFragment?
        if (valueFragment == null) {
            valueFragment = ValueFragment.newInstance("PopUp")
            fm.beginTransaction().replace(R.id.main, valueFragment!!, VALUE_TAG).commit()
        }



        if (NfcAdapter.ACTION_TECH_DISCOVERED == intent.action) {
            Log.i(TAG, "Started by tag discovery")
            onNewIntent(intent)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.popup_menu, menu)
        return true
    }

    public override fun onStart() {
        super.onStart()
        if (ValueHolder.getDataValues() != null)
            valueFragment!!.setValueData(ValueHolder.getDataValues() ?: return)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.fullscreen -> {
                val intent = Intent(this@PopupActivity, MainActivity::class.java)
                intent.action = getString(R.string.action_fullscreen_main)
                intent.putExtra(getString(R.string.extra_value_main), valueFragment!!.getValueData())

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    animateActivity21(intent)
                } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    animateActivity16(intent)
                } else {
                    startActivity(intent)
                    overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
                }

                return true
            }

            android.R.id.home ->{
                this.finish()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }


    @TargetApi(21)
    private fun animateActivity21(intent: Intent) {
        val options: ActivityOptions
        //TODO check what happens if last is empty
        if (valueFragment!!.getValueData() != null) {
            options = ActivityOptions.makeSceneTransitionAnimation(
                this@PopupActivity,
                Pair.create(findViewById(R.id.current), "current"),
                Pair.create(findViewById(R.id.last), "last"),
                Pair.create(findViewById(R.id.main_toolbar), "toolbar")
            )
        } else {
            options = ActivityOptions.makeSceneTransitionAnimation(
                this@PopupActivity,
                Pair.create(findViewById(R.id.current), "current"),
                Pair.create(findViewById(R.id.main_toolbar), "toolbar")
            )
        }

        startActivity(intent, options.toBundle())
        //ActivityCompat.startActivity(PopupActivity.this,intent,
        //		ActivityOptionsCompat.makeSceneTransitionAnimation(PopupActivity.this).toBundle());
        //finish();
    }

    @TargetApi(16)
    private fun animateActivity16(intent: Intent) {
        val root = findViewById<View>(R.id.popupRoot)

        val options = ActivityOptions.makeScaleUpAnimation(root, root.left, root.top, root.width, root.height)
        startActivity(intent, options.toBundle())
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        if (NfcAdapter.ACTION_TECH_DISCOVERED == intent.action) {
            Log.i(TAG, "Discovered tag with intent: $intent")
            val tag = intent.getParcelableExtra<Tag>(NfcAdapter.EXTRA_TAG)
            try {
                val `val` = Readers.getInstance().readTag(tag)
                valueFragment!!.setValueData(`val`)
            } catch (e: DesfireException) {
                Toast.makeText(this, R.string.communication_fail, Toast.LENGTH_SHORT).show()
            }catch (ist: IllegalStateException){
                Toast.makeText(this, R.string.communication_fail, Toast.LENGTH_SHORT).show()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        valueFragment = null
    }

    companion object {

        private val TAG = PopupActivity::class.java.simpleName

        private val VALUE_TAG = "value"
    }
}
